### ISSUE TITLE
PersistKeysToRedis not working in .NET Core 2.0 Preview 2 with Redis 0.1.2

### ISSUE THREAD

**Issue by varnav** [aspnet/DataProtection#251]
Hello! We're using Redis to store keys, but getting an error on startup:

> [Microsoft.AspNetCore.DataProtection.Repositories.EphemeralXmlRepository] - Using an in-memory repository. Keys will not be persisted to storage.
> [Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager] - Neither user profile nor HKLM registry available. Using an ephemeral key repository. Protected data will be unavailable when application exits.

Here is the code we use:

```
            string connect;
            try
            {
                var dnsTask = Dns.GetHostAddressesAsync(settings.RedisAddress);
                var addresses = dnsTask.Result;
                connect = string.Join(",", addresses.Select(x => x.MapToIPv4().ToString() + ":" + settings.RedisPort));
            }
            catch (Exception)
            {
                logger.LogError("Could not reslove host adress for Redis");
                throw new Exception("Could not reslove host adress for Redis");
            }

            var redis = ConnectionMultiplexer.Connect(connect);
            services.AddSingleton(options => redis.GetDatabase());

            services.AddDataProtection().PersistKeysToRedis(redis, "DataProtection-Keys");
```

**Comment by natemcmaster** [comment]
Can you provide more details? Can you share your Startup class or Program.cs file? Are you using the Microsoft.AspNetCore.All metapackage?

**Comment by akarimovbytewerk** [comment]
Hello, i'm from the same team as @varnav .
This is Program.cs:

`using System.IO;
using MboServer.Helpers;
using Microsoft.AspNetCore.Hosting;

namespace MboServer
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var helper = new SettingsHelper();
            var settings = helper.Load();

            var host = new WebHostBuilder()
                .UseKestrel()
                .UseUrls(settings.DefaultAddress)
                .UseContentRoot(Directory.GetCurrentDirectory())
                .UseStartup<Startup>()
                .Build();

            host.Run();
        }
    }
}
`

And Startup.cs :

`using System;
using System.Linq;
using System.Net;
using MboServer.Common;
using MboServer.DBLayer.ConnectionFactories;
using MboServer.DBLayer.Contexts;
using MboServer.DBLayer.Interfaces;
using MboServer.Extensions;
using MboServer.Helpers;
using Microsoft.AspNetCore.Authentication.Cookies;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.DataProtection;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.Net.Http.Headers;
using StackExchange.Redis;
using Swashbuckle.AspNetCore.Swagger;
using System.IO;
using MboServer.MessageSender;
using MboServer.Tools;
using Microsoft.AspNetCore.HttpOverrides;

namespace MboServer
{
    public class Startup
    {
        private readonly Settings settings;
        private readonly IHostingEnvironment enviroment;

        public Startup(IHostingEnvironment env)
        {
            var helper = new SettingsHelper();
            settings = helper.Load();
            enviroment = env;
        }

        /// <summary>
        /// This method gets called by the runtime. Use this method to add services to the container.
        /// </summary>
        /// <param name="services"></param>
        public void ConfigureServices(IServiceCollection services)
        {
            // Add framework services.
            services.AddMvc(
                config => {
                    config.Filters.Add(typeof(CustomExceptionFilter));
                })
                .AddXmlDataContractSerializerFormatters()
                .AddMvcOptions(opts => {
                    opts.FormatterMappings.SetMediaTypeMappingForFormat("xml", new MediaTypeHeaderValue("application/xml"));
                });

            var connectionFactory = new SqlConnectionFactory(settings.ConnectionString);
            services.AddSingleton<IConnectionFactory>(provider => connectionFactory);

            services.AddSingleton(provider => new StaticMethods());
            
            var loggerFactory = new LoggerFactory();
            loggerFactory.AddLog4Net();
            services.AddSingleton<ILoggerFactory>(loggerFactory);

            initRepositories(services);

            var logger = loggerFactory.CreateLogger("Engine");
            logger.LogInformation($"Connecting to Redis : {settings.RedisAddress}:{settings.RedisPort}");

            string connect;
            try
            {
                var dnsTask = Dns.GetHostAddressesAsync(settings.RedisAddress);
                var addresses = dnsTask.Result;
                connect = string.Join(",", addresses.Select(x => x.MapToIPv4().ToString() + ":" + settings.RedisPort));
            }
            catch (Exception)
            {
                logger.LogError("Could not reslove host adress for Redis");
                throw new Exception("Could not reslove host adress for Redis");
            }

            var redis = ConnectionMultiplexer.Connect(connect);
            services.AddSingleton(options => redis.GetDatabase());

            // Set key store to Redis
            services.AddDataProtection().PersistKeysToRedis(redis, "DataProtection-Keys");

            services.AddCookieAuthentication(o =>
            {
                o.LoginPath = new PathString("/api/User/Login");
                o.SlidingExpiration = true;
                o.ExpireTimeSpan = settings.SessionExpireTime;
                o.Events = new CookieAuthenticationEvents
                {
                    OnRedirectToLogin = CookieAuthenticationMethods.RedirectToLogin,
                    OnValidatePrincipal = CookieAuthenticationMethods.ValidateAsync
                };
                o.CookieSecure = enviroment.IsDevelopment() ? CookieSecurePolicy.SameAsRequest : CookieSecurePolicy.Always;
            });

            services.AddAuthentication(o =>
            {
                o.DefaultChallengeScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                o.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                o.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
            });

            // Register the Swagger generator, defining one or more Swagger documents
            services.AddSwaggerGen(c =>
            {
                c.SwaggerDoc("v1", new Info { Title = "Mbo API", Version = "v1" });

                //Set the comments path for the swagger json and ui.
                var basePath = Path.GetDirectoryName(System.Reflection.Assembly.GetEntryAssembly().Location);
                var xmlPath = Path.Combine(basePath, "MboServer.xml");
                c.IncludeXmlComments(xmlPath);
            });
        }

        private void initRepositories(IServiceCollection services)
        {
            
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory, IDatabase distributedCache)
        {
            app.UseStaticFiles(new StaticFileOptions
            {
                OnPrepareResponse = ctx =>
                {
                    ctx.Context.Response.Headers.Append("Cache-Control", "public,max-age=600,no-cache");
                }
            });

            CookieAuthenticationMethods.Cache = distributedCache;
            app.UseAuthentication();

            app.UseForwardedHeaders(new ForwardedHeadersOptions
            {
                ForwardedHeaders = ForwardedHeaders.XForwardedFor |ForwardedHeaders.XForwardedProto
            });

            var logger = loggerFactory.CreateLogger("Engine");
            logger.LogInformation("Server is started successfully");

            app.UseMvc();

            // Enable middleware to serve generated Swagger as a JSON endpoint.
            app.UseSwagger();

            // Enable middleware to serve swagger-ui (HTML, JS, CSS etc.), specifying the Swagger JSON endpoint.
            app.UseSwaggerUI(c =>
            {
                c.RoutePrefix = "metadata";
                c.SwaggerEndpoint("/swagger/v1/swagger.json", "Mbo API V1");
            });            
        }
    }
}
`

**Comment by akarimovbytewerk** [comment]
No we don't using Microsoft.AspNetCore.All
Redis is included via NuGet manager as 'StackExchange.Redis.StrongName' library

**Comment by natemcmaster** [comment]
Are you using Microsoft.AspNetCore.DataProtection.Redis 0.1.2?

**Comment by akarimovbytewerk** [comment]
Yes as described in official Mocrosoft recomendation https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/implementation/key-storage-providers
The reason is that i need to store net core protection keys in Redis (method PersistKeysToRedis).
I would be glad to hear another solution.

**Comment by akarimovbytewerk** [comment]
P.S. It was working fine on net core 1.1

**Comment by natemcmaster** [comment]
Microsoft.AspNetCore.DataProtection.Redis 0.1.2 was designed to work with ASP.NET Core 1.x. Since then, we have refactored the way that `PersistKeysToRedis` ties into the main data protection services. However, we didn't publish the updated version to NuGet.org. This was an oversight, and we hope will be fixed soon. In the meantime, you can work around this in one of the following ways:

1. (Recommended) Manually add the redis repository to `KeyManagementOptions`. Example:
    ```c#
    public void ConfigureServices(IServiceCollection services)
    {
        var redis = ConnectionMultiplexer.Connect("localhost:6379");
        services.AddDataProtection().PersistKeysToRedis(redis);
        services.Configure<KeyManagementOptions>(o =>
        {
            o.XmlRepository = new RedisXmlRepository(() => redis.GetDatabase(), "DataProtection-Keys");
        });
    }
    ```

2. Use an updated package from the nightly feeds. This build is closest to what is on the `rel/2.0.0-preview2` branch of this repo: [Microsoft.AspNetCore.DataProtection.Redis 0.3.0-preview2-25794 on MyGet.org](https://dotnet.myget.org/feed/aspnetcore-release/package/nuget/Microsoft.AspNetCore.DataProtection.Redis/0.3.0-preview2-25794)

FYI - the source of the bug. Compare these:
[RedisDataProtectionBuilderExtensions](https://github.com/aspnet/DataProtection/blob/rel/1.1.2/src/Microsoft.AspNetCore.DataProtection.Redis/RedisDataProtectionBuilderExtensions.cs#L71) from the rel/1.1.2 branch
[RedisDataProtectionBuilderExtensions](https://github.com/aspnet/DataProtection/blob/rel/2.0.0-preview2/src/Microsoft.AspNetCore.DataProtection.Redis/RedisDataProtectionBuilderExtensions.cs#L73) from the rel/2.0.0-preview2 branch.

**Comment by akarimovbytewerk** [comment]
Using you suggestion 
```
public void ConfigureServices(IServiceCollection services)
{
    var redis = ConnectionMultiplexer.Connect("localhost:6379");
    services.AddDataProtection();
    services.Configure<KeyManagementOptions>(o =>
    {
        o.XmlRepository = new RedisXmlRepository(() => redis.GetDatabase(), "DataProtection-Keys");
    });
}
```
the key was saved successfully but there is new error in log:
[Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider] - An error occurred while reading the key ring

**Comment by natemcmaster** [comment]
I've [updated](https://github.com/aspnet/Coherence/pull/235) our build infrastructure so the 0.3.0 version of Microsoft.AspNetCore.DataProtection.Redis will be published when 2.0.0 RTM ships. Until then, use the workaround mentioned above or a nightly version from our pre-release feeds.

Closing as there are no more action items on our end. Users should update to 0.3.0 when the package is available on NuGet.org.
